core/ir/synthetic/encoding_common.h

```diff
@@ -0,0 +1,155 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#ifndef _SYNTHETIC_ENCODING_COMMON_H_
+#define _SYNTHETIC_ENCODING_COMMON_H_
+
+/**
+ * This synthetic ISA (which we call DR_ISA_REGDEPS) is a made up ISA that has the purpose
+ * of preserving register dependencies and giving hints on the type of operations each
+ * instruction is performing.
+ * For this reason the majority of operations that would normally work on instructions
+ * coming from an actual ISA (e.g., DR_ISA_AMD64) are not supported.
+ * We support instr_convert_to_isa_regdeps(), which converts an #instr_t of an actual ISA
+ * to a DR_ISA_REGDEPS instruction.
+ * Other operations we support (related to encoding and decoding of a DR_ISA_REGDEPS
+ * #instr_t) are: instr_encode(), instr_encode_to_copy(), decode(), decode_common().
+ * DR_ISA_REGDEPS #instr_t can only return the encoded information described in the
+ * encoding scheme (core/ir/synthetic/encoding_common.h): categories (from #dr_category_t)
+ * to indicate the type of operation performed (e.g., a load, a math operation, floating
+ * point math operation, a branch, etc.), arithmetic flags, number of source and
+ * destination register operands, number of memory operations (i.e., loads + stores)
+ * performed (stored in #instr_t encoding_hints field), register operands and their size
+ * (which is the same for all operands), instruction length, and ISA mode (which is
+ * DR_ISA_REGDEPS).
+ * Querying additional information (e.g., the instruction opcode) will return the default
+ * value generated by instr_create() (i.e., zero).
+ */
+
+/* XXX i#1684: Note that this is part of a larger issue, where lack of cross-arch support
+ * in the same build of DynamoRIO is limiting us.
+ */
+
+/* Here we describe the encoding scheme for the Synthetic ISA that is enforced in decode.c
+ * and encode.c.
+ *
+ * Encoded instructions are 4 byte aligned.
+ *
+ * All instruction encodings begin with the following 5 header bytes, which follow this
+ * scheme:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *        category         eflags #src   #dst
+ * |--------| |--------| |--------| |--------|
+ * 31.. ..24  23.. ..16  15..  ..8  7..   ..0
+ *  padding    padding    padding    #mem_ops
```

Comment:
We need 1 byte for #mem_ops because I miscalculated the number of bits available in op_size (we need 6 bits for op_size, leaving 2 for #mem_ops and I wasn't sure it was enough, although we can reduce the number of bits for op_size if we do a re-map of the enum values, given that we only care about register sizes, I just didn't think it was worth the additional complexity).
Also, #mem_ops is now always part of the header because we can't rely on #src and #dst of register operands (like we do for op_size to make it optional) since we can have instructions with 0 register operands that perform memory operations in x86 (e.g., MOV r/m32, imm32).
If we are willing to sacrifice 1 bit in the category to signal the presence of memory operations, we can make #mem_ops optional.

Reply:
Maybe best to just bail on the # mem ops if it's not easy.  It might also be confusing if someone thinks it will include the dynamic data-dependent count for scatter/gather ops.

Reply:
Agreed.
Done, reverted back to no #mem_ops, header is back to 4 bytes.
Updated doc in core/ir/synthetic/encoding_common.h and PR description.